src/app/qgsdatasourcemanagerdialog.cpp

```diff
@@ -0,0 +1,170 @@
+/***************************************************************************
+    qgsdatasourcemanagerdialog.cpp - datasource manager dialog
+
+    ---------------------
+    begin                : May 19, 2017
+    copyright            : (C) 2017 by Alessandro Pasotti
+    email                : apasotti at itopen dot it
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include <QMessageBox>
+#include <QListWidgetItem>
+
+#include "qgsdatasourcemanagerdialog.h"
+#include "ui_qgsdatasourcemanagerdialog.h"
+#include "qgsbrowserdockwidget.h"
+#include "qgssettings.h"
+#include "qgsproviderregistry.h"
+#include "qgsopenvectorlayerdialog.h"
+#include "qgssourceselectdialog.h"
+#include "qgsmapcanvas.h"
+
+
+QgsDataSourceManagerDialog::QgsDataSourceManagerDialog( QgsMapCanvas *mapCanvas, QWidget *parent, Qt::WindowFlags fl ) :
+  QgsOptionsDialogBase( QStringLiteral( "Data Source Manager" ), parent, fl ),
+  ui( new Ui::QgsDataSourceManagerDialog ),
+  mMapCanvas( mapCanvas )
+{
+
+  ui->setupUi( this );
+  // QgsOptionsDialogBase handles saving/restoring of geometry, splitter and current tab states,
+  // switching vertical tabs between icon/text to icon-only modes (splitter collapsed to left),
+  // and connecting QDialogButtonBox's accepted/rejected signals to dialog's accept/reject slots
+  initOptionsBase( true );
+
+  // Bind list index to the stacked dialogs
+  connect( ui->mOptionsListWidget, SIGNAL( currentRowChanged( int ) ), this, SLOT( setCurrentPage( int ) ) );
+
+  // BROWSER Add the browser widget to the first stacked widget page
+  mBrowserWidget = new QgsBrowserDockWidget( QStringLiteral( "Browser" ), this );
+  mBrowserWidget->setFeatures( QDockWidget::NoDockWidgetFeatures );
+  ui->mOptionsStackedWidget->addWidget( mBrowserWidget );
+
+  // VECTOR Layers (completely different interface: it's not a provider)
+  QgsOpenVectorLayerDialog *ovl = new QgsOpenVectorLayerDialog( this, Qt::Widget, true );
+  ui->mOptionsStackedWidget->addWidget( ovl );
+  QListWidgetItem *ogrItem = new QListWidgetItem( tr( "Vector files" ), ui->mOptionsListWidget );
+  ogrItem->setIcon( QgsApplication::getThemeIcon( QStringLiteral( "/mActionAddOgrLayer.svg" ) ) );
+  connect( ovl, &QgsOpenVectorLayerDialog::addVectorLayers, this, &QgsDataSourceManagerDialog::vectorLayersAdded );
+
+  // Add data provider dialogs
+  QDialog *dlg;
+
+#ifdef HAVE_POSTGRESQL
```

Comment:
Could we avoid all this hardcoding and just get the providers direct from the registry?